rotkehlchen/chain/ethereum/tokens.py

```diff
@@ -141,116 +172,65 @@ def query_tokens_for_addresses(
+        current_time = ts_now()
+        for account in accounts:
+            if self.ethereum.connected_to_any_web3():
+                querying_etherscan = False
+                tokens_batch_size = OTHER_MAX_TOKEN_CHUNK_LENGTH
+            else:
+                querying_etherscan = True
+                tokens_batch_size = ETHERSCAN_MAX_TOKEN_CHUNK_LENGTH
+            token_balances = self._query_chunks(
+                account=account,
+                chunks=get_chunks(all_tokens, n=tokens_batch_size),
+                querying_etherscan=querying_etherscan,
+            )
+            detected_tokens = list(token_balances.keys())
+            accounts_info[account] = (detected_tokens, current_time)
+            self.db.save_tokens_for_address(write_cursor, account, detected_tokens)
+        return accounts_info
 
-    def _get_multitoken_account_balance(
+    def query_tokens_for_addresses(
             self,
-            tokens: List[EthereumToken],
-            account: ChecksumEthAddress,
-            call_order: Optional[Sequence[WeightedNode]],
-    ) -> Dict[str, FVal]:
-        """Queries balances of multiple tokens for an account
-
-        Return a dictionary with keys being tokens and value a dictionary of
-        account to balances
+            addresses: List[ChecksumEthAddress],
+    ) -> TokenBalancesType:
+        """Queries token balances for a list of addresses
+        Returns the token balances of each address and the usd prices of the tokens
 
         May raise:
         - RemoteError if an external service such as Etherscan is queried and
           there is a problem with its query.
         - BadFunctionCallOutput if a local node is used and the contract for the
           token has no code. That means the chain is not synced
         """
-        log.debug(
-            'Querying ethereum chain for multi token account balances',
-            eth_address=account,
-            tokens_num=len(tokens),
-        )
-        balances: Dict[str, FVal] = {}
-        result = ETH_SCAN.call(
-            ethereum=self.ethereum,
-            method_name='tokensBalance',
-            arguments=[account, [x.ethereum_address for x in tokens]],
-            call_order=call_order,
-        )
-        for tk_idx, token in enumerate(tokens):
-            token_amount = result[tk_idx]
-            if token_amount != 0:
-                normalized_amount = token_normalized_value(token_amount, token)
-                log.debug(
-                    f'Found {token.symbol}({token.ethereum_address}) token balance for '
-                    f'{account} and amount {normalized_amount}',
-                )
-                balances[token.identifier] = normalized_amount
-        return balances
+        address_to_balance: Dict[ChecksumEthAddress, Dict[EthereumToken, FVal]] = {}
+        all_tokens = set()
+
+        for address in addresses:
+            with self.db.conn.read_ctx() as cursor:
```

Comment:
Move the db cursor context manager out of the loop. Why would you recreate a cursor and a context manager for each address and loop iteration?

Reply:
Makes sense